source/non_matching/fe_values.cc

```diff
@@ -0,0 +1,399 @@
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#include <deal.II/base/exceptions.h>
+
+#include <deal.II/lac/block_vector.h>
+#include <deal.II/lac/la_parallel_block_vector.h>
+#include <deal.II/lac/la_parallel_vector.h>
+#include <deal.II/lac/la_vector.h>
+#include <deal.II/lac/petsc_block_vector.h>
+#include <deal.II/lac/petsc_vector.h>
+#include <deal.II/lac/trilinos_epetra_vector.h>
+#include <deal.II/lac/trilinos_parallel_block_vector.h>
+#include <deal.II/lac/trilinos_tpetra_vector.h>
+#include <deal.II/lac/trilinos_vector.h>
+#include <deal.II/lac/vector.h>
+
+#include <deal.II/non_matching/fe_values.h>
+
+#include <deal.II/numerics/fe_field_function.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace NonMatching
+{
+  RegionUpdateFlags::RegionUpdateFlags()
+    : inside(update_default)
+    , outside(update_default)
+    , surface(update_default)
+  {}
+
+
+  namespace internal
+  {
+    namespace FEValuesImplementation
+    {
+      template <int dim, class VectorType>
+      RefSpaceFEFieldFunction<dim, VectorType>::RefSpaceFEFieldFunction(
+        const DoFHandler<dim> &dof_handler,
+        const VectorType &     dof_values)
+        : dof_handler(&dof_handler)
+        , global_dof_values(&dof_values)
+      {
+        Assert(dof_handler.n_dofs() == dof_values.size(),
+               ExcDimensionMismatch(dof_handler.n_dofs(), dof_values.size()));
+      }
+
+
+
+      template <int dim, class VectorType>
+      void
+      RefSpaceFEFieldFunction<dim, VectorType>::set_active_cell(
+        const typename Triangulation<dim>::active_cell_iterator &cell)
+      {
+        Assert(
+          &cell->get_triangulation() == &dof_handler->get_triangulation(),
+          ExcMessage(
+            "The incoming cell must belong to the triangulation associated with "
+            "the DoFHandler passed to the constructor."));
+
+        const typename DoFHandler<dim>::active_cell_iterator dof_handler_cell(
+          &dof_handler->get_triangulation(),
+          cell->level(),
+          cell->index(),
+          dof_handler);
+
+        // Save the element and the local dof values, since this is what we need
+        // to evaluate the function.
+        element = &dof_handler_cell->get_fe();
+
+        local_dof_indices.resize(element->dofs_per_cell);
+        dof_handler_cell->get_dof_indices(local_dof_indices);
+
+        local_dof_values.resize(element->dofs_per_cell);
+
+        for (unsigned int i = 0; i < local_dof_indices.size(); i++)
+          local_dof_values[i] =
+            dealii::internal::ElementAccess<VectorType>::get(
+              *global_dof_values, local_dof_indices[i]);
```

Comment:
Why not use a dummy `AffineConstraints` object?

Reply:
I tried this:
```c++

const AffineConstraints<> dummy_constraints;
dummy_constraints.get_dof_values(*global_dof_values,
                                 local_dof_indices.begin(),
                                 local_dof_values.begin(),
                                 local_dof_values.end());
```
But it doesn't seem to work for `EpetraWrappers::Vector`

```
2021-11-19T15:40:38.4878272Z /home/runner/work/dealii/dealii/include/deal.II/lac/affine_constraints.h: In instantiation of ‘void dealii::AffineConstraints<number>::get_dof_values(const VectorType&, ForwardIteratorInd, ForwardIteratorVec, ForwardIteratorVec) const [with ForwardIteratorVec = __gnu_cxx::__normal_iterator<double*, std::vector<double, std::allocator<double> > >; ForwardIteratorInd = __gnu_cxx::__normal_iterator<unsigned int*, std::vector<unsigned int> >; VectorType = dealii::LinearAlgebra::EpetraWrappers::Vector; number = double]’:
2021-11-19T15:40:38.4945499Z /home/runner/work/dealii/dealii/source/non_matching/fe_values.cc:96:9:   required from ‘void dealii::NonMatching::internal::FEValuesImplementation::RefSpaceFEFieldFunction<dim, VectorType>::set_active_cell(const typename dealii::Triangulation<dim>::active_cell_iterator&) [with int dim = 1; VectorType = dealii::LinearAlgebra::EpetraWrappers::Vector; typename dealii::Triangulation<dim>::active_cell_iterator = dealii::TriaActiveIterator<dealii::CellAccessor<1, 1> >]’
2021-11-19T15:40:38.4948301Z /home/runner/work/dealii/dealii/build/source/non_matching/fe_values.inst:991:2:   required from here
2021-11-19T15:40:38.4950239Z /home/runner/work/dealii/dealii/include/deal.II/lac/affine_constraints.h:2337:44: error: no match for call to ‘(const dealii::LinearAlgebra::EpetraWrappers::Vector) (unsigned int&)’
2021-11-19T15:40:38.4951503Z          *local_vector_begin = global_vector(*local_indices_begin);
2021-11-19T15:40:38.4952226Z                                ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
2021-11-19T15:40:38.4953966Z /home/runner/work/dealii/dealii/include/deal.II/lac/affine_constraints.h:2344:36: error: no match for call to ‘(const dealii::LinearAlgebra::EpetraWrappers::Vector) (const unsigned int&)’
2021-11-19T15:40:38.4955245Z              value += (global_vector(position.entries[j].first) *
2021-11-19T15:40:38.4955955Z                        ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~
```